### ISSUE TITLE
Wrong shape painter color

### ISSUE THREAD

**Issue by PascalLadalle** [4ian/GDevelop#2644]
I'm not sure how to explain this bug...
The fill color is changing in cycles incrementing by 1 each cycle. (color change after 1 click, then color change after 2 clicks, then after 3 clicks...).

![4a01ec81a2d749cd6ecc9ae82660b29acc4d23fd](https://user-images.githubusercontent.com/25978830/118114702-3a7d3200-b3e8-11eb-9a42-ee9a650350dd.gif)

[Bug_ShapePainter.zip](https://github.com/4ian/GDevelop/files/6471918/Bug_ShapePainter.zip)

reported on [the forum](https://forum.gdevelop-app.com/t/problem-with-changing-fill-color-of-a-shape-painter/32569/6)

**Comment by 4ian** [comment]
The GIF seems not really exactly what is happening in the project, but I can confirm this is really weird. I've quickly checked but the fill color seems properly to stay the same internally.. a Pixi bug? Might have to dive more to check what's going on.

**Comment by Bouh** [comment]
The fill color and outline color from the shape painter are wrong.
whe value shoud be on format **0x000000**, but there is just **255**
I don't know why the rgb to hex number return a weird value...

**Comment by Bouh** [comment]
At first frame it's ok
![image](https://user-images.githubusercontent.com/1670670/118122464-2db20b80-b3f3-11eb-91a4-98df3ecf7366.png)

When we click, it's bugged
![image](https://user-images.githubusercontent.com/1670670/118122483-36a2dd00-b3f3-11eb-8f76-a7ca00137eed.png)

**Comment by Bouh** [comment]
What happend :o 

![image](https://user-images.githubusercontent.com/1670670/118122781-98fbdd80-b3f3-11eb-9534-4f0cc6a542d0.png)

**Comment by 4ian** [comment]
How do you get the fill color to change? 

I've added a console.log:
![image](https://user-images.githubusercontent.com/1280130/118123102-aad98280-b3eb-11eb-94d6-e45f3a86668c.png)

and it's always the same value for me: 
![image](https://user-images.githubusercontent.com/1280130/118123135-bb89f880-b3eb-11eb-984b-ddcb3a914c7f.png)

Note that 255 *is* a valid value.
All blue color: 0x0000FF = 0xFF = 255 
All red color: 0xFF0000 = 16711680

**Comment by PascalLadalle** [comment]
I tried again from scratch, just in case, and I confirm the issue.
Few tests I did:
- replaced cursor condition by timer condition: no change
- removed the initial fill color action: no change
- removed the initial draw action: fixed
- replaced the "at the beginning" by a timer condition: bug doesn't occur for the first 5 seconds (even though debug text shows the wrong colour should be displayed), then occurs more or less at each timer reset.
- replaced "at the beginning" by a manual trigger: bug doesn't occur until manually triggered, then keeps occurring without further triggers, but not in the same way as before, the cycle thing doesn't seem valid anymore.

So there's somehow a conflict between the two drawing actions, even when the first one is behind a false condition.

[Bug_ShapePainter.zip](https://github.com/4ian/GDevelop/files/6472654/Bug_ShapePainter.zip)

**Comment by Bouh** [comment]
It's the same in Pixi 5.3.3 
https://game-previews.gdevelop-app.com/1620913880543-685093/index.html

As the button **Clear the rendered image between each frame** isn't checked all previous graphics are still in memory in an array on pixi side.
But the new shape should be on top with the colors from the object which is a correct value at each frame.

**Comment by Bouh** [comment]
I've added console log in Pixi, in graphics data, batching, and where is called fillStyle, I see always the right color.

For example a breakpoint here
![image](https://user-images.githubusercontent.com/1670670/118145080-7c6c9f00-b40d-11eb-987a-a2bc240ddc81.png)

With this loop on graphic called from the beginning:
```
// this = shape painter runtime object
this.getRendererObject().geometry.graphicsData.forEach(
    (item)=>{
        console.log(item.fillStyle.color);
    }
)
```

![image](https://user-images.githubusercontent.com/1670670/118145137-90b09c00-b40d-11eb-9e08-1b59bfa4ace3.png)

As the first frame is red the value is 16711680 which is correct.
All the next frames are blue which is also correct, but during these frames, the circle has been red twice! 
And there is no change in graphicsData.

I really don't know what happend and where I need to search now..

**Comment by tristanbob** [comment]
I wonder if this bug is related to the strange color behavior on this game example:
https://victrisgames.itch.io/demo-of-travel-to-random-positions

(Follow this exact order of steps to trigger bug)
1) Click "Show Path Details"
2) Set "Zoom" to 0.2
3) Set "Move Speed" to 1000
4) Set "Direction Bias" to 0.5
5) Watch the circles be drawn with rainbow color gradients instead of solid colors

https://user-images.githubusercontent.com/8879811/118148044-6b03a280-b3cd-11eb-8c86-576f7fdee2bb.mp4

**Comment by Bouh** [comment]
I've found how isolate the bug, when the shape painter action are called once the bug happend.
**At the beginning of the scene** and **Trigger once** have a side effect on actions.

There is probably a problem in the order of the calls to the Pixi.Graphics

**Comment by Bouh** [comment]
The bug happend when the graphics are never cleared.

A workaround could be make a render of the shape painter texture, use it as texture, clear all graphics in pixi batch renderer, and loop like this for each frames or if graphics call is too high.

I've seen in `GDevelop\GDJS\Runtime\pixi-renderers\layer-pixi-renderer.ts` a function called `_updateRenderTexture` that can be interesting.

I've started to write something similar for shape painter, I've not yet the time to try it.

**Comment by 4ian** [comment]
That's a workaround, but this also comes with other implications (higher memory usage, need to generate a texture). Ideally we would understand why it's "semi randomly" failing 🤔

**Comment by Bouh** [comment]
> That's a workaround, but this also comes with other implications (higher memory usage, need to generate a texture). Ideally we would understand why it's "semi randomly" failing 🤔

I've successfully isolate the bug in a Pixi playground.
https://codesandbox.io/s/pixijs-bounding-box-forked-81xwd?file=/src/index.js

**Comment by 4ian** [comment]
Awesome! Looks definitely like a Pixi issue then? Do you want to report it on their repository?

**Comment by Bouh** [comment]
Probably fixed by https://github.com/pixijs/pixijs/pull/7529
Wait for a new build of PixiJS on v6 branch.

**Comment by 4ian** [comment]
Thanks for following that up 👍👍

**Comment by ClementPasteau** [comment]
Tested this example with Pixi.js 6.1.2 and the bug is fixed :+1: 
See PR #3026 
We'll do a few tests to make sure there isn't any regression and then we can merge it for next release

**Comment by ClementPasteau** [comment]
Fixed by #3026 ! This will be available in the next release.
